'Cannot delegate to an abstract method' for scoped bean with complex interfaces

## Describe the bug
For a scoped bean with the setup:

`MyBean extends Component implements HasSize` constructor calls `HasSize::setSize`
interface `HasSize extends HasElement` has a default implementation `setSize` that calls `HasElement::getElement`
interface `HasElement`
class `Component implements HasElement` This has the actual implementation

### Expected behavior
Bean can be instantiated and injected.

### Actual behavior
Bean instantiation throws `Cannot delegate to an abstract method`

```
Caused by: java.lang.IllegalStateException: Cannot delegate to an abstract method
        at org.acme.MyBean_ClientProxy.getElement(MyBean_ClientProxy.zig:229)
        at org.acme.HasSize.setSize(HasSize.java:7)
        at org.acme.MyBean_ClientProxy.setSize(MyBean_ClientProxy.zig:168)
        at org.acme.MyBean.<init>(MyBean.java:11)
        at org.acme.MyBean_ClientProxy.<init>(MyBean_ClientProxy.zig:28)
        at org.acme.MyBean_Bean.proxy(MyBean_Bean.zig:37)
        at org.acme.MyBean_Bean.get(MyBean_Bean.zig:217)
        at org.acme.MyBean_Bean.get(MyBean_Bean.zig:233)
        at org.acme.GreetingResource_Bean.create(GreetingResource_Bean.zig:154)
        ... 37 more
```

## To Reproduce

https://github.com/caalador/quarkus-scoped-bean

Or attach an archive containing the reproducer to the issue.

Steps to reproduce the behavior:
1. run `mvn clean package -Dquarkus.debug.generated-sources-dir=gen`
2. see that test fails
3. Search `gen/org/acme/MyBean_ClientProxy.zig` for `Method getElement` and see that it has `IllegalStateException` for if bean not instantiated

## Environment (please complete the following information):
windows 10

### Output of `java -version`
Java version: 1.8.0_222 & Java version: 11.0.4

### Quarkus version or git rev
1.10.5.Final, 1.13.6.Final, 2.0.0.CR3

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)
Apache Maven 3.8.1

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ClientProxyGenerator.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ClientProxyGenerator.java
@@ -33,6 +33,7 @@ import java.util.function.Predicate;
 import org.jboss.jandex.ClassInfo;
 import org.jboss.jandex.DotName;
 import org.jboss.jandex.FieldInfo;
+import org.jboss.jandex.IndexView;
 import org.jboss.jandex.MethodInfo;
 import org.jboss.jandex.Type;
 import org.jboss.jandex.TypeVariable;
@@ -310,11 +311,12 @@ public class ClientProxyGenerator extends AbstractGenerator {
     Collection<MethodInfo> getDelegatingMethods(BeanInfo bean, Consumer<BytecodeTransformer> bytecodeTransformerConsumer,
             boolean transformUnproxyableClasses) {
         Map<Methods.MethodKey, MethodInfo> methods = new HashMap<>();
+        IndexView index = bean.getDeployment().getBeanArchiveIndex();
 
         if (bean.isClassBean()) {
             Set<Methods.NameAndDescriptor> methodsFromWhichToRemoveFinal = new HashSet<>();
             ClassInfo classInfo = bean.getTarget().get().asClass();
-            Methods.addDelegatingMethods(bean.getDeployment().getBeanArchiveIndex(), classInfo,
+            Methods.addDelegatingMethods(index, classInfo,
                     methods, methodsFromWhichToRemoveFinal, transformUnproxyableClasses);
             if (!methodsFromWhichToRemoveFinal.isEmpty()) {
                 String className = classInfo.name().toString();
@@ -323,16 +325,16 @@ public class ClientProxyGenerator extends AbstractGenerator {
             }
         } else if (bean.isProducerMethod()) {
             MethodInfo producerMethod = bean.getTarget().get().asMethod();
-            ClassInfo returnTypeClass = getClassByName(bean.getDeployment().getBeanArchiveIndex(), producerMethod.returnType());
-            Methods.addDelegatingMethods(bean.getDeployment().getBeanArchiveIndex(), returnTypeClass, methods, null,
+            ClassInfo returnTypeClass = getClassByName(index, producerMethod.returnType());
+            Methods.addDelegatingMethods(index, returnTypeClass, methods, null,
                     transformUnproxyableClasses);
         } else if (bean.isProducerField()) {
             FieldInfo producerField = bean.getTarget().get().asField();
-            ClassInfo fieldClass = getClassByName(bean.getDeployment().getBeanArchiveIndex(), producerField.type());
-            Methods.addDelegatingMethods(bean.getDeployment().getBeanArchiveIndex(), fieldClass, methods, null,
+            ClassInfo fieldClass = getClassByName(index, producerField.type());
+            Methods.addDelegatingMethods(index, fieldClass, methods, null,
                     transformUnproxyableClasses);
         } else if (bean.isSynthetic()) {
-            Methods.addDelegatingMethods(bean.getDeployment().getBeanArchiveIndex(), bean.getImplClazz(), methods, null,
+            Methods.addDelegatingMethods(index, bean.getImplClazz(), methods, null,
                     transformUnproxyableClasses);
         }
 

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Methods.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Methods.java
@@ -69,8 +69,8 @@ final class Methods {
 
     static void addDelegatingMethods(IndexView index, ClassInfo classInfo, Map<MethodKey, MethodInfo> methods,
             Set<NameAndDescriptor> methodsFromWhichToRemoveFinal, boolean transformUnproxyableClasses) {
-        // TODO support interfaces default methods
         if (classInfo != null) {
+            // First methods declared on the class
             for (MethodInfo method : classInfo.methods()) {
                 if (skipForClientProxy(method, transformUnproxyableClasses, methodsFromWhichToRemoveFinal)) {
                     continue;
@@ -88,18 +88,20 @@ final class Methods {
                             key.method.exceptions().toArray(Type.EMPTY_ARRAY));
                 });
             }
-            // Interfaces
-            for (Type interfaceType : classInfo.interfaceTypes()) {
-                ClassInfo interfaceClassInfo = getClassByName(index, interfaceType.name());
-                if (interfaceClassInfo != null) {
-                    addDelegatingMethods(index, interfaceClassInfo, methods, methodsFromWhichToRemoveFinal,
-                            transformUnproxyableClasses);
-                }
-            }
+            // Methods declared on superclasses
             if (classInfo.superClassType() != null) {
                 ClassInfo superClassInfo = getClassByName(index, classInfo.superName());
                 if (superClassInfo != null) {
                     addDelegatingMethods(index, superClassInfo, methods, methodsFromWhichToRemoveFinal,
+                            transformUnproxyableClasses);
+                }
+            }
+            // Methods declared on implemented interfaces
+            // TODO support interfaces default methods
+            for (DotName interfaceName : classInfo.interfaceNames()) {
+                ClassInfo interfaceClassInfo = getClassByName(index, interfaceName);
+                if (interfaceClassInfo != null) {
+                    addDelegatingMethods(index, interfaceClassInfo, methods, methodsFromWhichToRemoveFinal,
                             transformUnproxyableClasses);
                 }
             }

--- a/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/clientproxy/delegatingmethods/ClientProxyMethodInvocationInConstructorTest.java
+++ b/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/clientproxy/delegatingmethods/ClientProxyMethodInvocationInConstructorTest.java
@@ -1,0 +1,25 @@
+package io.quarkus.arc.test.clientproxy.delegatingmethods;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import io.quarkus.arc.Arc;
+import io.quarkus.arc.ClientProxy;
+import io.quarkus.arc.test.ArcTestContainer;
+import java.io.IOException;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+public class ClientProxyMethodInvocationInConstructorTest {
+
+    @RegisterExtension
+    public ArcTestContainer container = new ArcTestContainer(MyBean.class, Component.class, HasElement.class, HasSize.class);
+
+    @Test
+    public void testClientProxy() throws IOException {
+        // Just test that the client proxy can be instantiated
+        MyBean myBean = Arc.container().instance(MyBean.class).get();
+        assertTrue(myBean instanceof ClientProxy);
+        assertEquals("an element", myBean.getElement().toString());
+    }
+}

--- a/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/clientproxy/delegatingmethods/Component.java
+++ b/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/clientproxy/delegatingmethods/Component.java
@@ -1,0 +1,11 @@
+package io.quarkus.arc.test.clientproxy.delegatingmethods;
+
+public class Component implements HasElement {
+
+    private static final long serialVersionUID = 1L;
+
+    @Override
+    public Object getElement() {
+        return "an element";
+    }
+}

--- a/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/clientproxy/delegatingmethods/HasElement.java
+++ b/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/clientproxy/delegatingmethods/HasElement.java
@@ -1,0 +1,10 @@
+package io.quarkus.arc.test.clientproxy.delegatingmethods;
+
+import java.io.Serializable;
+
+@FunctionalInterface
+public interface HasElement extends Serializable {
+
+    Object getElement();
+
+}

--- a/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/clientproxy/delegatingmethods/HasSize.java
+++ b/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/clientproxy/delegatingmethods/HasSize.java
@@ -1,0 +1,8 @@
+package io.quarkus.arc.test.clientproxy.delegatingmethods;
+
+public interface HasSize extends HasElement {
+
+    default void setSize(String size) {
+        getElement();
+    }
+}

--- a/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/clientproxy/delegatingmethods/MyBean.java
+++ b/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/clientproxy/delegatingmethods/MyBean.java
@@ -1,0 +1,14 @@
+package io.quarkus.arc.test.clientproxy.delegatingmethods;
+
+import javax.enterprise.context.ApplicationScoped;
+
+@ApplicationScoped
+public class MyBean extends Component implements HasSize {
+
+    private static final long serialVersionUID = 1L;
+
+    public MyBean() {
+        setSize("5px");
+    }
+
+}

